components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1269,7 +1270,131 @@ where
             SignificantMsg::RaftlogFetched { context, res } => {
                 self.on_raft_log_fetched(context, res);
             }
+            SignificantMsg::EnterForceLeaderState { failed_stores } => {
+                self.on_enter_pre_force_leader(failed_stores)
+            }
+            SignificantMsg::ExitForceLeaderState => self.on_exit_force_leader(),
+        }
+    }
+
+    fn on_enter_pre_force_leader(&mut self, failed_stores: HashSet<u64>) {
+        if self.fsm.peer.force_leader.is_some() {
+            return;
+        }
+
+        let expected_alive_voter = self
+            .fsm
+            .peer
+            .raft_group
+            .raft
+            .prs()
+            .conf()
+            .voters()
+            .ids()
+            .iter()
+            .filter(|v| !failed_stores.contains(v))
+            .collect();
+        info!(
+            "enter pre force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+            "alive_voter" => ?expected_alive_voter,
+        );
+
+        if self.fsm.peer.is_leader() {
+            // expire the lease of previous leader.
+            self.fsm.peer.leader_lease.expire();
+            self.fsm
+                .peer
+                .raft_group
+                .raft
+                .become_follower(self.fsm.peer.term(), raft::INVALID_ID);
+        }
+
+        // become candidate first to increase term
+        self.fsm.peer.raft_group.raft.become_candidate();
```

Comment:
How about sending itself `MsgHup`?

Reply:
actually, `campaign()` does this thing. Here calling `become_candidate` is to increase the term which would be easy for us to notice the change in logs if we are investigating a problem.